test/dotnet.Tests/CommandTests/ToolManifestTests.cs

```diff
@@ -0,0 +1,183 @@
+
+            manifestResult.ShouldBeEquivalentTo(_defaultExpectedResult);
+        }
+
+        [Fact]
+        public void GivenManifestFileOnParentDirectoryItGetContent()
+        {
+            var subdirectoryOfTestRoot = Path.Combine(_testDirectoryRoot, "sub");
+            _fileSystem.File.WriteAllText(Path.Combine(_testDirectoryRoot, _manifestFilename), _jsonContent);
+            var toolManifest = new ToolManifestFinder(new DirectoryPath(subdirectoryOfTestRoot), _fileSystem);
+            var manifestResult = toolManifest.Find();
+
+            manifestResult.ShouldBeEquivalentTo(_defaultExpectedResult);
+        }
+
+        [Fact]
+        // https://github.com/JamesNK/Newtonsoft.Json/issues/931#issuecomment-224104005
+        // Due to a limitation of newtonsoft json
+        public void GivenManifestWithDuplicatedPackageIdItReturnsTheLastValue()
+        {
+            _fileSystem.File.WriteAllText(Path.Combine(_testDirectoryRoot, _manifestFilename),
+                _jsonWithDuplicatedPackagedId);
+            var toolManifest = new ToolManifestFinder(new DirectoryPath(_testDirectoryRoot), _fileSystem);
+            var manifestResult = toolManifest.Find();
+
+            manifestResult.Should()
+                .Contain(
+                    new ToolManifestFindingResultSinglePackage(
+                        new PackageId("t-rex"),
+                        NuGetVersion.Parse("2.1.4"),
+                        new[] { new ToolCommandName("t-rex") }));
+        }
+
+        [Fact]
+        public void WhenCalledWithFilePathItGetContent()
+        {
+            string customFileName = "customname.file";
+            _fileSystem.File.WriteAllText(Path.Combine(_testDirectoryRoot, customFileName), _jsonContent);
+            var toolManifest = new ToolManifestFinder(new DirectoryPath(_testDirectoryRoot), _fileSystem);
+            var manifestResult =
+                toolManifest.Find(new FilePath(Path.Combine(_testDirectoryRoot, customFileName)));
+
+            manifestResult.ShouldBeEquivalentTo(_defaultExpectedResult);
+        }
+
+        [Fact]
+        public void WhenCalledWithNonExistsFilePathItThrows()
+        {
+            var toolManifest = new ToolManifestFinder(new DirectoryPath(_testDirectoryRoot), _fileSystem);
+            Action a = () => toolManifest.Find(new FilePath(Path.Combine(_testDirectoryRoot, "non-exits")));
+            a.ShouldThrow<ToolManifestException>().And.Message.Should().Contain(string.Format(LocalizableStrings.CannotFindAnyManifestsFileSearched, ""));
+        }
+
+        [Fact]
+        public void GivenNoManifestFileItThrows()
+        {
+            var toolManifest = new ToolManifestFinder(new DirectoryPath(_testDirectoryRoot), _fileSystem);
+            Action a = () => toolManifest.Find();
+            a.ShouldThrow<ToolManifestException>().And.Message.Should().Contain(string.Format(LocalizableStrings.CannotFindAnyManifestsFileSearched, ""));
```

Comment:
Pending discussion

Reply:
Maybe we are thinking in different scenarios. If the command cannot be found, it should throw, no?

Reply:
If I type `dotnet asdf` then I should get an error message if `asdf` doesn't resolve to anything.  But the error should be the same regardless of whether I have no local tools installed, or whether I just don't have any with an `asdf` command.

Reply:
I see. The existing resolver's "protocol" is to return null when cannot resolve command. And it will move forward to the next resolver. And in the end, a generic "cannot find" message will be printed out.

However i don't think I will reuse "toolManifest.Find" to resolve the command. toolManifest.Find will find all manifest entry and return a combined list. It will read all possible json file. During resolve command, it should only return one entry and it should stop looking for more if found.